File: packages/nerva-py/nerva/registry/inmemory.py
"""In-memory registry — for testing and simple deployments."""

from __future__ import annotations

from typing import TYPE_CHECKING

from nerva.registry import (
    ComponentKind,
    HealthStatus,
    RegistryEntry,
    RegistryPatch,
)

if TYPE_CHECKING:
    from nerva.context import ExecContext


# Fields on RegistryEntry that RegistryPatch can overwrite.
_PATCHABLE_FIELDS = (
    "description",
    "metadata",
    "health",
    "enabled",
    "requirements",
    "permissions",
)


class InMemoryRegistry:
    """Registry backed by a plain dict. No persistence.

    Suitable for tests and single-process deployments where component
    registration does not need to survive restarts.

    ``discover()`` filters results by:
    - ``kind`` matches the requested component type
    - ``enabled`` is ``True``
    - ``health`` is not ``UNAVAILABLE``
    - If the entry declares ``permissions``, the caller must hold at
      least one matching role in ``ctx.permissions.roles``
    """

    def __init__(self) -> None:
        self._entries: dict[str, RegistryEntry] = {}

    async def register(self, entry: RegistryEntry, ctx: ExecContext) -> None:
        """Add or replace a component in the registry.

        If an entry with the same ``name`` already exists, it is overwritten.

        Args:
            entry: Component definition to register.
            ctx: Execution context carrying identity and permissions.
        """
        self._entries[entry.name] = entry

    async def discover(
        self, kind: ComponentKind, ctx: ExecContext
    ) -> list[RegistryEntry]:
        """List components of a given kind visible to the caller.

        Filters out disabled entries, unavailable entries, and entries
        whose required permissions are not satisfied by ``ctx.permissions``.

        Args:
            kind: Component type to filter by.
            ctx: Execution context used for permission checks.

        Returns:
            List of matching ``RegistryEntry`` objects, sorted by name.
        """
        results: list[RegistryEntry] = []
        for entry in self._entries.values():
            if not _matches_discovery_criteria(entry, kind, ctx):
                continue
            results.append(entry)

        return sorted(results, key=lambda e: e.name)

    async def resolve(
        self, name: str, ctx: ExecContext
    ) -> RegistryEntry | None:
        """Look up a single component by name.

        Args:
            name: Unique component identifier.
            ctx: Execution context (reserved for future permission gating).

        Returns:
            The matching ``RegistryEntry``, or ``None`` if not found.
        """
        return self._entries.get(name)

    async def health(self, name: str) -> HealthStatus:
        """Get the current health status of a component.

        Args:
            name: Unique component identifier.

        Returns:
            Current ``HealthStatus``.

        Raises:
            KeyError: If no component with the given name is registered.
        """
        entry = self._entries.get(name)
        if entry is None:
            raise KeyError(f"Component not found: {name!r}")
        return entry.health

    async def update(self, name: str, patch: RegistryPatch) -> None:
        """Apply a partial update to a registered component.

        Only non-``None`` fields in the patch are written to the entry.

        Args:
            name: Unique component identifier to update.
            patch: Fields to overwrite.

        Raises:
            KeyError: If no component with the given name is registered.
        """
        entry = self._entries.get(name)
        if entry is None:
            raise KeyError(f"Component not found: {name!r}")

        _apply_patch(entry, patch)


def _matches_discovery_criteria(
    entry: RegistryEntry, kind: ComponentKind, ctx: ExecContext
) -> bool:
    """Check whether a registry entry passes all discovery filters.

    Args:
        entry: Candidate entry to evaluate.
        kind: Required component type.
        ctx: Execution context carrying caller permissions.

    Returns:
        ``True`` if the entry should be included in discovery results.
    """
    if entry.kind != kind:
        return False
    if not entry.enabled:
        return False
    if entry.health == HealthStatus.UNAVAILABLE:
        return False
    if entry.permissions and not _has_required_permission(entry, ctx):
        return False
    return True


def _has_required_permission(entry: RegistryEntry, ctx: ExecContext) -> bool:
    """Check that the caller holds at least one role required by the entry.

    Args:
        entry: Entry with a non-empty permissions list.
        ctx: Execution context carrying caller roles.

    Returns:
        ``True`` if the caller has at least one matching role.
    """
    caller_roles = ctx.permissions.roles
    return bool(caller_roles & frozenset(entry.permissions))


def _apply_patch(entry: RegistryEntry, patch: RegistryPatch) -> None:
    """Write non-None patch fields onto the entry.

    Args:
        entry: Entry to mutate in place.
        patch: Partial update — only non-None fields are applied.
    """
    for field_name in _PATCHABLE_FIELDS:
        value = getattr(patch, field_name)
        if value is not None:
            setattr(entry, field_name, value)

File: packages/nerva-js/src/registry/inmemory.ts
/**
 * In-memory registry — for testing and simple deployments.
 *
 * Backed by a plain `Map`. No persistence across restarts.
 *
 * @module registry/inmemory
 */

import type {
  ExecContext,
  RegistryEntry,
  RegistryPatch,
} from "./index.js";
import {
  ComponentKind,
  HealthStatus,
} from "./index.js";

// ---------------------------------------------------------------------------
// Patchable fields
// ---------------------------------------------------------------------------

/** Fields on RegistryEntry that RegistryPatch can overwrite. */
const PATCHABLE_FIELDS = [
  "description",
  "metadata",
  "health",
  "enabled",
  "requirements",
  "permissions",
] as const;

type PatchableField = typeof PATCHABLE_FIELDS[number];

// ---------------------------------------------------------------------------
// InMemoryRegistry
// ---------------------------------------------------------------------------

/**
 * Registry backed by a plain Map. No persistence.
 *
 * Suitable for tests and single-process deployments where component
 * registration does not need to survive restarts.
 *
 * `discover()` filters results by:
 * - `kind` matches the requested component type
 * - `enabled` is true
 * - `health` is not UNAVAILABLE
 * - If the entry declares `permissions`, the caller must hold at
 *   least one matching role in `ctx.permissions.roles`
 */
export class InMemoryRegistry {
  private readonly entries: Map<string, RegistryEntry> = new Map();

  /**
   * Add or replace a component in the registry.
   *
   * If an entry with the same name already exists, it is overwritten.
   *
   * @param entry - Component definition to register.
   * @param _ctx - Execution context (reserved for future use).
   */
  async register(entry: RegistryEntry, _ctx: ExecContext): Promise<void> {
    this.entries.set(entry.name, entry);
  }

  /**
   * List components of a given kind visible to the caller.
   *
   * Filters out disabled entries, unavailable entries, and entries
   * whose required permissions are not satisfied by `ctx.permissions`.
   *
   * @param kind - Component type to filter by.
   * @param ctx - Execution context used for permission checks.
   * @returns List of matching RegistryEntry objects, sorted by name.
   */
  async discover(kind: ComponentKind, ctx: ExecContext): Promise<RegistryEntry[]> {
    const results: RegistryEntry[] = [];

    for (const entry of this.entries.values()) {
      if (matchesDiscoveryCriteria(entry, kind, ctx)) {
        results.push(entry);
      }
    }

    return results.sort((a, b) => a.name.localeCompare(b.name));
  }

  /**
   * Look up a single component by name.
   *
   * @param name - Unique component identifier.
   * @param _ctx - Execution context (reserved for future permission gating).
   * @returns The matching RegistryEntry, or null if not found.
   */
  async resolve(name: string, _ctx: ExecContext): Promise<RegistryEntry | null> {
    return this.entries.get(name) ?? null;
  }

  /**
   * Get the current health status of a component.
   *
   * @param name - Unique component identifier.
   * @returns Current HealthStatus.
   * @throws {@link Error} If no component with the given name is registered.
   */
  async health(name: string): Promise<HealthStatus> {
    const entry = this.entries.get(name);
    if (!entry) {
      throw new Error(`Component not found: "${name}"`);
    }
    return entry.health;
  }

  /**
   * Apply a partial update to a registered component.
   *
   * Only non-undefined fields in the patch are written to the entry.
   *
   * @param name - Unique component identifier to update.
   * @param patch - Fields to overwrite.
   * @throws {@link Error} If no component with the given name is registered.
   */
  async update(name: string, patch: RegistryPatch): Promise<void> {
    const entry = this.entries.get(name);
    if (!entry) {
      throw new Error(`Component not found: "${name}"`);
    }
    applyPatch(entry, patch);
  }
}

// ---------------------------------------------------------------------------
// Pure helpers
// ---------------------------------------------------------------------------

/**
 * Check whether a registry entry passes all discovery filters.
 *
 * @param entry - Candidate entry to evaluate.
 * @param kind - Required component type.
 * @param ctx - Execution context carrying caller permissions.
 * @returns True if the entry should be included in discovery results.
 */
function matchesDiscoveryCriteria(
  entry: RegistryEntry,
  kind: ComponentKind,
  ctx: ExecContext,
): boolean {
  if (entry.kind !== kind) {
    return false;
  }
  if (!entry.enabled) {
    return false;
  }
  if (entry.health === HealthStatus.UNAVAILABLE) {
    return false;
  }
  if (entry.permissions.length > 0 && !hasRequiredPermission(entry, ctx)) {
    return false;
  }
  return true;
}

/**
 * Check that the caller holds at least one role required by the entry.
 *
 * @param entry - Entry with a non-empty permissions list.
 * @param ctx - Execution context carrying caller roles.
 * @returns True if the caller has at least one matching role.
 */
function hasRequiredPermission(entry: RegistryEntry, ctx: ExecContext): boolean {
  const callerRoles = ctx.permissions.roles;
  for (const requiredRole of entry.permissions) {
    if (callerRoles.has(requiredRole)) {
      return true;
    }
  }
  return false;
}

/**
 * Write non-undefined patch fields onto the entry.
 *
 * @param entry - Entry to mutate in place.
 * @param patch - Partial update — only non-undefined fields are applied.
 */
function applyPatch(entry: RegistryEntry, patch: RegistryPatch): void {
  for (const fieldName of PATCHABLE_FIELDS) {
    const value = patch[fieldName as PatchableField];
    if (value !== undefined) {
      // Safe to assign: fieldName is a mutable field on RegistryEntry
      // and the patch value type matches.
      (entry as unknown as Record<string, unknown>)[fieldName] = value;
    }
  }
}
